Video player (2 of 2): adding closed captioning

clrux opened the pull request:
This pull request is the second of two parts and adds closed-captioning to our video player. It is part of [AC-188](https://openedx.atlassian.net/browse/AC-188).

## Notable changes

* adds a "CC" button to the grouped button container with the transcript and language control
* relies on Draggabilly, which is already included and loaded in the platform
* sets a cookie so the captioning option persists through video viewing across the platform
* allows placement of the captions within the video container for personalization

## Browsers

This has been tested in the following browsers:

| Platform | Browser | Version | Passed |
|----------|---------|---------|--------|
| Mac  | Chrome | 45  | Yes  |
|          | Safari | 8  | Yes |
|          | Firefox | 40 | Yes |
| Windows | Chrome  | 45 | Yes |
|          | Firefox | 40 | Yes |
|          | IE | 10 | Yes |
|          | IE | 11 | Yes |

## Sandbox

https://clrux-ac-188.sandbox.edx.org/courses/course-v1:edX+AV101+2015_T3/courseware/C_01/lecture_01/

## Reviewers

- [x] Frances (FED and Sass)
- [ ] Mark (Accessibility)
- [ ] Christine (JS + Tests)
- [ ] Carol (Documentation)
- [ ] T&L review (JS + Tests)

clrux commented:
paver test a11y

clrux commented:
paver test accessibility

clrux commented:
jenkins run a11y

clrux commented:
jenkins run js

clrux commented:
jenkins run bokchoy

clrux commented:
jenkins run lettuce

clrux commented:
jenkins run a11y
jenkins run quality
jenkins run bokchoy
jenkins run js

clrux commented:
jenkins run lettuce

clrux commented:
jenkins run a11y

cahrens commented:
@clrux When do you expect this PR to be ready for TNL review (we'd like a full first review done first)?

clrux commented:
@cahrens We discovered a few small accessibility things on the first part of this work right before I left for vacation. I was hoping my team could review this work while I was away to minimize the hold up, but they didn't get around to it. Hold off for now. I'm going to review the first part and get it merged first then I'll re-up this one for you guys once my team reviews. I'll let @explorerleslie know as well.

clytwynec commented on common/lib/xmodule/xmodule/js/spec/video/video_caption_spec.js:
```diff
@@ -158,6 +171,54 @@
+                        };
+
+                    it('toggle the captions on control click', function() {
+                        state = jasmine.initializePlayer();
```
It looks like in the "on click" test the original state of the captions is _visible_, but in the "on keypress space/enter" tests the original state is _not visible_.   Why is that different?

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -1,11 +1,14 @@
-    function (Sjson, AsyncProcess) {
+    ['video/00_sjson.js', 'video/00_async_process.js', 'draggabilly'],
+    function (Sjson, AsyncProcess, Draggabilly) {
+        // TODO: remove once the Jasmine tests are able to load Draggabilly correctly
```
This TODO should be addressed before merging.

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -149,10 +171,14 @@
                         mouseenter: this.onContainerMouseEnter,
-                        mouseleave: this.onContainerMouseLeave
-                    });
+                        mouseleave: this.onContainerMouseLeave,
```
Why this change?

clytwynec commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -226,13 +226,28 @@ def test_cc_button_wo_english_transcript(self):
-    def test_cc_button_transcripts_and_sub_fields_empty(self):
+    def test_cc_button(self):
+        """
+        Scenario: CC button works correctly with transcript in YouTube move
```
typo? ... I think it's supposed to say  "YouTube mode" instead of "YouTube move".

clytwynec commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -226,13 +226,28 @@ def test_cc_button_wo_english_transcript(self):
+    def test_cc_button(self):
+        """
+        Scenario: CC button works correctly with transcript in YouTube move
+        Given the course has a vieo component in "Youtube" mode
```
typo: "vieo"

clytwynec commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -226,13 +226,28 @@ def test_cc_button_wo_english_transcript(self):
+        self.navigate_to_video()
+        self.video.show_closed_captions()
+
+        self.assertTrue(self.video.is_closed_captions_visible)
```
This will probably need a "wait" to make sure this isn't flaky. It looks like you could probably use `self.video.wait_for_closed_captions`.

Could you also add a couple lines after this to make sure that hiding the captions works. (`self.video.hide_closed_captions()` then check that the captions are no longer visible)

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -1017,6 +1052,98 @@
                 }
             },
 
+            handleCaptioningCookie: function() {
```
Is the `.cookie` method part of jquery by default or does it need to be "required"?

clrux replied:
There was already cookie functions in this file and I just copied and modified them, so I'm assuming the jQuery plugin is already loaded. It stores the cookie properly.

clytwynec commented:
Ok, I just finished my first review of this.  Looks pretty good, but it looks like there are a couple of gaps in automated tests.  In particular, I don't see any tests for: 
* the new cookie for captioning
* changing languages with the cc button to check that it does change language for captions and that it correctly applies that language to transcripts too.

clrux commented:
jenkins run quality
jenkins run lettuce
jenkins run a11y

clrux commented:
jenkins run bokchoy
jenkins run lettuce
jenkins run python

clrux commented:
jenkins run quality

clrux commented:
jenkins run quality

clytwynec commented on common/test/acceptance/pages/lms/video/video.py:
```diff
@@ -293,6 +295,18 @@ def hide_captions(self):
+        """
+        self._closed_captions_visibility(False)
+
     def is_captions_visible(self):
```
This is actually referring to transcripts right?

clrux replied:
Correct.

clytwynec commented on common/test/acceptance/pages/lms/video/video.py:
```diff
@@ -327,7 +353,26 @@ def _captions_visibility(self, captions_new_state):
+        # Make sure that the captions are visible
+        EmptyPromise(lambda: self.is_closed_captions_visible() == closed_captions_new_state,
+                     "Closed captions are {state}".format(state=state)).fulfill()
+        # EmptyPromise(self.is_closed_captions_visible,
```
Commented out code is still here.

clytwynec commented on common/test/acceptance/pages/lms/video/video.py:
```diff
@@ -346,6 +391,22 @@ def captions_text(self):
+        self.wait_for_closed_captions()
+
+        closed_captions_selector = self.get_element_selector(CSS_CLASS_NAMES['closed_captions'])
+        subs = self.q(css=closed_captions_selector).html
```
What does "subs" stand for?  "substring"?

clrux replied:
This test was duplicated from the one above for the transcripts. I believe `subs` here refers to the text, either in the transcript or in the caption (whichever the test is checking).

clytwynec commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -554,6 +589,46 @@ def test_video_language_menu_working(self):
+        # Change the language to Chinese and check for correct text
+        self.video.select_language('zh')
+        self.video.click_player_button('play')
+        self.video.wait_for_position('0:05')
```
Does this need to wait a whole 5 seconds before pausing?  If not, it is just adding an extra 5 seconds to every test run.

clrux replied:
Testing locally needs a different amount of time than testing on Jenkins. I was using 5 to be safe, but I could probably try 2 or 3.

clytwynec replied:
Is it needed to ensure that the video has started playing before pausing it? If that's the case, it looks like you could use `self.video.wait_for_state('playing')`.

clrux replied:
Not sure, but don't think so. The video plays, it just plays choppy (at least locally), so sometimes 00:01 doesn't have the text, but other times it might. I chose a time that should guarantee text.

clytwynec commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -554,6 +589,46 @@ def test_video_language_menu_working(self):
+        # Change the language to English and check for correct text
+        self.video.select_language('en')
+        self.video.click_player_button('play')
+        self.video.wait_for_position('0:05')
```
Same here... does it need to wait that long?

clrux replied:
See above note.

cptvitamin commented:
@clrux this is looking good.  I few things I noticed:  

1. There seems to be a correlation with dragabilly.js and moment.js returning 404's and the video hanging at load (spinning icon).  We talked about video hanging and I think you told me it was limited to devstack.  I just wanted to bring it up and let you know I found that correlation.
2. If you attempt to turn on the captions before starting the video, the first caption loads, and then disappears, leaving the empty caption box on screen.  It would be great if this didn't happen at all (disappearing caption), and if it can't be helped at all, it would be great if the window had a zero height.
3. The background for the caption bar is much too transparent.  Could you swap the hover transparency level with the default?
4. If the user has enabled "Always show captions" in their Youtube profile, the Youtube captions compete with the edX captions.  There doesn't seem to be a way to detect if the user has this setting enabled.  This is more likely something we will have to cover in documentation.  If automatic captions are available on Youtube and this setting is turned on, the user will most likely want to consume the edX caption stream but will not be able to turn off the automated ones from LMS.  Very few users know they have to actually go to Youtube to change this setting.  Another way we could solve this issue is by removing the transparency from our caption box (hiding/covering the youtube captions)

clytwynec commented:
@clrux I just merged PR #10671 to fix the jshint diff reporting.  Can you rebase this off master to get an accurate diff-quality report?

clrux commented:
@cptvitamin Thank for the feedback. I am not sure what to do about the RequireJS stuff. Moment was something that was already there and being loaded, and I copied its code to get Draggabilly to load. The pathing for Moment did not change. I'm not sure why we get 404's on local. It doesn't happen on sandies.

I've addressed number two above by determining if there is `.current` text in the transcript and if not, using placeholder text instead so the user knows text will appear there once the video plays.

I've also darkened the default state of the closed captions.

Finally, I've added the parameter to the YouTube video iframe, but I don't think it'll work. See [this StackOverflow](http://stackoverflow.com/questions/25803804/disable-youtube-captions-on-embedded-video) about there being only one value (`1`).

@clytwynec Thank you. I've rebased and updated this branch, along with the sandy.

@andy-armstrong When running this locally, we sometimes get 404's with `moment.js` and `Draggabilly`, and it looks like the path doesn't always load properly. This will cause the video not to finish loading. It's random so it makes me think it may be some sort of race condition, or something isn't completing on the build before loading the video. It doesn't seem to have issues on sandboxes, but any insights here would be greatly appreciated.

cahrens commented:
We have been seeing issues with moment ever since it was added to the video module's requirejs configuration. It doesn't surprise me that the same issue would be seen with draggabilly. I think that this will need to be figured out (addressing both libraries) before another dependency is added (caching will also be an issue).
 
Here is the bug we have for moment-- https://openedx.atlassian.net/browse/TNL-3311. I am not sure if we can work on this bug in our next iteration, but we can discuss it at planning today.

clrux commented:
@cahrens @andy-armstrong Can either of you lend me a hand with these two JS tests? I've spent a total of three days now trying to get them to pass, but can't figure it out. Any help would be appreciated.

cc @clytwynec

cahrens commented:
@clrux is your branch rebased on master after the 1.8 upgrade (not that it would cause an issue with these JS tests, but I just want to make sure before trying to run the branch)?

clrux commented:
@cahrens It has been rebased. I'm trying one more thing right now and I'll commit that work and notes for you to see what attempts I've made. I'll ping you in a little while if this last test doesn't work.
